zapcore/buffered_write_syncer.go

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapcore
+
+import (
+	"bufio"
+	"sync"
+	"time"
+)
+
+// A BufferedWriteSyncer is a WriteSyncer that can also flush any buffered data
+// with the ability to change the buffer size, flush interval and Clock.
+// The default values are; Size 256kb, FlushInterval 30s.
+type BufferedWriteSyncer struct {
+	WriteSyncer
+
+	Size          int
+	FlushInterval time.Duration
+	Clock         Clock
+
+	// unexported fields for state
+	mu          sync.Mutex
+	writer      *bufio.Writer
+	ticker      *time.Ticker
+	stop        chan struct{}
+	initialized bool
+}
+
+const (
+	// _defaultBufferSize specifies the default size used by Buffer.
+	_defaultBufferSize = 256 * 1024 // 256 kB
+
+	// _defaultFlushInterval specifies the default flush interval for
+	// Buffer.
+	_defaultFlushInterval = 30 * time.Second
+)
+
+func (s *BufferedWriteSyncer) loadConfig() {
+	size := s.Size
+	if size == 0 {
+		size = _defaultBufferSize
+	}
+
+	flushInterval := s.FlushInterval
+	if flushInterval == 0 {
+		flushInterval = _defaultFlushInterval
+	}
+
+	if s.Clock == nil {
+		s.Clock = DefaultClock
+	}
+	s.ticker = s.Clock.NewTicker(flushInterval)
+
+	s.writer = bufio.NewWriterSize(s.WriteSyncer, size)
+	s.stop = make(chan struct{})
+	s.initialized = true
+	go s.flushLoop()
```

Comment:
dunno, i recall having discussions with @abhinav about this before, but i think i prefer if go routines start not in loadConfig but below, in Write function. However, I may be wrong here.

Reply:
In this case, loadConfig is like our constructor. Compare the following two:

```
func NewFoo(...) *Foo { /* initialize internals */ }

func (*Foo) DoThing() { ... }

// versus

type Foo struct{}

func (*Foo) loadConfig() { /* initialize internals */ }

func (f *Foo) DoThing() {
  if !f.loaded { f.loadConfig() }
  ...
}
```

So loadConfig is effectively doing the job of the constructor, in which case it makes sense to spawn the goroutine there because all initialization should happen in the same place. (Spawning the goroutine is part of initialization here because there will only ever be one goroutine per instance of this object, tied to that object. If we were spawning goroutines on a per-request basis, then that would be in Write.)